common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java

```diff
@@ -22,64 +22,311 @@
     private static final String TAG = ClientCertAuthChallengeHandler.class.getSimpleName();
     private static final String ACCEPTABLE_ISSUER = "CN=MS-Organization-Access";
     private Activity mActivity;
     private final YubiKitManager mYubiKitManager;
+    private UsbYubiKeyDevice mDevice;
+
+    //creating nested class to hold details of a Certificate needed for the picker,
+    // including subject, issuer, and slot.
+    public class YubiKitCertDetails implements Serializable {
+        private final String issuerText;
+        private final String subjectText;
+        private final Slot slot;
+
+        public YubiKitCertDetails(String issuerText, String subjectText, Slot slot) {
+            this.issuerText = issuerText;
+            this.subjectText = subjectText;
+            this.slot = slot;
+        }
+
+        public String getIssuerText() {
+            return issuerText;
+        }
+
+        public String getSubjectText() {
+            return subjectText;
+        }
+
+        public Slot getSlot() {
+            return slot;
+        }
+    }
 
     public ClientCertAuthChallengeHandler(@NonNull final Activity activity) {
         mActivity = activity;
-        //Create and start YubiKitManager
+        //set mDevice to null
+        mDevice = null;
+        //Create and start YubiKitManager for UsbDiscovery mode.
+        //When in Usb Discovery mode, Yubikeys that plug into the device will be accessible
+        // once the user provides permission via the Android permission dialog.
         mYubiKitManager = new YubiKitManager(mActivity.getApplicationContext());
         mYubiKitManager.startUsbDiscovery(new UsbConfiguration(), new Callback<UsbYubiKeyDevice>() {
             @Override
             public void invoke(UsbYubiKeyDevice device) {
                 Logger.info(TAG, "A YubiKey device was connected");
-                device.setOnClosed(new Runnable() {
+                //set device
+                mDevice = device;
+                mDevice.setOnClosed(new Runnable() {
                     @Override
                     public void run() {
                         Logger.info(TAG, "A YubiKey device was disconnected");
+                        mDevice = null;
                     }
                 });
             }
         });
     }
 
+    //Called when a ClientCertRequest is received by the WebViewClient.
+    //Offers the user to choose a certificate to authenticate with based on whether or not a YubiKey is plugged in and has permission to be connected.
     @TargetApi(Build.VERSION_CODES.LOLLIPOP)
     @Override
     public Void processChallenge(@NonNull final ClientCertRequest request) {
-        //final String methodTag = TAG + ":processChallenge";
-        return handleOnDeviceCertAuth(request);
+        final String methodTag = TAG + ":processChallenge";
+        // Check if a YubiKey device has been discovered and has permission to be connected.
+        if (mDevice != null) {
```

Comment:
If this is concerning the `if` statement, in which way could I use ResultFuture instead of synchronized here? The part that complicates it for me is the `else` statement, as if there isn't a device plugged in, we want to show the other on-device certs instead.

Reply:
Update: Tried using ResultFuture to wait for results from the cert picker (this implementation would have the overall same functionality, but I wouldn't have to separate the listeners into separate methods. This would be cleaner.) However, I noticed when I unplugged while the cert picker was showing, the thread was blocked by resultFuture.get(), despite me setting resultFuture.setResult(null) in the mDevice.setOnClosed() callback. This confuses me a bit, since I thought the setOnClosed() callback was running on a separate thread, but it seems that I may need to experiment more with ResultFuture if we want to take that route.